rules/private/diff_test_private.bzl

```diff
@@ -0,0 +1,170 @@
+        else:
+            was_same = "echo>\"%DIFF_TEST_OUT%\" @echo Expected %DIFF_TEST_SRC1% and %DIFF_TEST_SRC2% to be different & echo>>\"%DIFF_TEST_OUT%\" @exit /B 1"
+            was_different = "echo>\"%DIFF_TEST_OUT%\" @exit /B 0"
+
+        # Write the source code of 'diff_script'.
+        ctx.actions.write(
+            output = diff_script,
+            content = content.format(was_same = was_same, was_different = was_different),
+            is_executable = True,
+        )
+
+        # Run 'diff_script' to generate 'test_script'.
+        ctx.actions.run(
+            inputs = [ctx.file.file1, ctx.file.file2, diff_script],
+            outputs = [test_script],
+            executable = "cmd.exe",
+            arguments = ["/C", diff_script.path.replace("/", "\\")],
+            env = {
+                "DIFF_TEST_SRC1": ctx.file.file1.path.replace("/", "\\"),
+                "DIFF_TEST_SRC2": ctx.file.file2.path.replace("/", "\\"),
+                "DIFF_TEST_OUT": test_script.path.replace("/", "\\"),
+            },
+            progress_message = "Comparing files",
+            mnemonic = "CompareFiles",
+        )
+    else:
+        test_script = ctx.actions.declare_file(ctx.label.name + "-test.sh")
+
+        # Source of the shell command that generates 'test_script'.
+        #
+        # This command runs at build time and compares the two files using
+        # "diff". Then it writes a shell script ('test_script') which is the
+        # actual script executed by "bazel test".
+        #
+        # If the files are the same and were expected to be so, or if they are
+        # different and were expected to be so, then the command writes
+        # 'test_script' to do nothing and exit with zero. Otherwise the command
+        # writes 'test_script' to print an error and exit with 1.
+        #
+        # If one of the files was not a regular file (or it was a symlink whose
+        # target is not a regular file) then this shell command fails.
+        command = """set -euo pipefail
+if ! test -f "$1"; then echo "File not found: $1" >&2 ; exit 1; fi
+if ! test -f "$2"; then echo "File not found: $2" >&2 ; exit 1; fi
+if diff -q "$1" "$2" >/dev/null 2>&1; then echo -e {was_same} > "$3"; else echo -e {was_different} > "$3"; fi
+chmod +x "$3"
+"""
+
+        # Contents of 'test_script' depending on diff result and expectation.
+        if ctx.attr.expect_same:
+            was_same = "'#!/bin/sh\nexit 0'"
+            was_different = "\"#!/bin/sh\necho 'Expected \'$1\' and \'$2\' to be the same' >&2\nexit 1\""
+        else:
+            was_same = "\"#!/bin/sh\necho 'Expected \'$1\' and \'$2\' to be different' >&2\nexit 1\""
+            was_different = "'#!/bin/sh\nexit 0'"
+
+        # Run the shell command to generate 'test_script'.
+        ctx.actions.run_shell(
```

Comment:
The code is a bit surprising. You run the diff as part of the build, and you generate a dummy file (exit 0 / exit 1) that will be executed during the test.

Instead, you could write the shell script with `actions.write_file`. Then the `diff` is executed as part of the test.
(the same pattern as in https://github.com/bazelbuild/examples/blob/master/rules/test_rule/line_length.bzl)